content/blog/2023-10-18-skywalking-toolkit-trace/index.md

```diff
@@ -0,0 +1,285 @@
+	"github.com/apache/skywalking-go/toolkit/trace"
+)
+
+func getProvider() (*http.Response, error) {
+	// Create an HTTP request
+	req, err := http.NewRequest("GET", "http://localhost:9998/provider", http.NoBody)
+	// Create an ExitSpan before sending the HTTP request.
+	trace.CreateExitSpan("GET:/provider", "localhost:9999",
+		func(headerKey, headerValue string) error {
+			// Injector adds specific header information to the request.
+			req.Header.Add(headerKey, headerValue)
+			return nil
+		})
+	// Finish the ExitSpan and ensure it executes when the function returns using defer.
+	defer trace.StopSpan()
+
+	// Send the request.
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func consumerHandler(w http.ResponseWriter, r *http.Request) {
+	// Create an EntrySpan to trace the execution of the consumerHandler method.
+	trace.CreateEntrySpan(r.Method+"/consumer", func(headerKey string) (string, error) {
+		// Extractor retrieves the header information added to the request.
+		return r.Header.Get(headerKey), nil
+	})
+	// Finish the EntrySpan.
+	defer trace.StopSpan()
+
+	// Prepare to send an HTTP request.
+	resp, err := getProvider()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return
+	}
+	_, _ = w.Write(body)
+}
+
+func main() {
+	http.HandleFunc("/consumer", consumerHandler)
+
+	_ = http.ListenAndServe(":9999", nil)
+}
+```
+
+```go
+// provider.go
+package main
+
+import (
+	"net/http"
+
+	_ "github.com/apache/skywalking-go"
+	"github.com/apache/skywalking-go/toolkit/trace"
+)
+
+func providerHandler(w http.ResponseWriter, r *http.Request) {
+	//Create an EntrySpan to trace the execution of the providerHandler method.
+	trace.CreateEntrySpan("GET:/provider", func(headerKey string) (string, error) {
+		return r.Header.Get(headerKey), nil
+	})
+	// Finish the EntrySpan.
+	defer trace.StopSpan()
+
+	_, _ = w.Write([]byte("success from provider"))
+}
+
+func main() {
+	http.HandleFunc("/provider", providerHandler)
+
+	_ = http.ListenAndServe(":9998", nil)
+}
+
+```
+
+
+
+Then, in the terminal, execute:
+
+```shell
+go build -toolexec="/path/to/go-agent" -a -o consumer ./consumer.go
+./consumer
+```
+
+```shell
+go build -toolexec="/path/to/go-agent" -a -o provider ./provider.go
+./provider
+```
+
+```shell
+curl 127.0.0.1:9999/consumer
+```
+
+At this point, the UI will display the span information you created.
+
+![Trace List](./trace.png)
+
+If you need to trace methods that are executed only locally, you can use trace.CreateLocalSpan(). If you don't need to monitor information or states from the other end, you can change ExitSpan and EntrySpan to LocalSpan.
+
+The usage examples provided are for illustration purposes, and users can decide the tracing granularity and where in the program they need tracing.
+
+Please note that if a program ends too quickly, it may cause tracing data to be unable to be asynchronously sent to the SkyWalking backend.
```

Comment:
I am not sure if it is possible but can a manual sync be forced for short running programs?

Reply:
Currently, skywalking-go does not support the feature of forced synchronization.

Reply:
Sebastian asked a similar question on the slack channel too in PoC codes.

And I agree, we should keep async as always, and no plan to add that.